### ISSUE TITLE
`IO` during task insertion causes race condition on `DAG`?

### ISSUE THREAD

**Issue by maltezfaria** [maltezfaria/DataFlowTasks.jl#16]
I have noticed in the cholesky factorization that if we ever fall back to the *generic* version of the `memory_overlap` (because for example we used the adjoint of a matrix but did not define a specific `memory_overlap` method for that), the program crashes with the following error:

```julia
┌ Warning: memory_overlap(::Matrix{Float64},::Adjoint{Float64, Matrix{Float64}}) not implemented.
└ @ DataFlowTasks ~/Projects/DataFlowTasks/src/dataflowtask.jl:154
ERROR: KeyError: key DataFlowTask (done) 230 not found
Stacktrace:
  [1] getindex
    @ ~/.julia/packages/OrderedCollections/PRayh/src/ordered_dict.jl:380 [inlined]
  [2] outneighbors
    @ ~/Projects/DataFlowTasks/src/dag.jl:89 [inlined]
  [3] addedge!(dag::DataFlowTasks.DAG{DataFlowTasks.DataFlowTask}, i::DataFlowTasks.DataFlowTask, j::DataFlowTasks.DataFlowTask)
    @ DataFlowTasks ~/Projects/DataFlowTasks/src/dag.jl:105
  [4] update_edges!(dag::DataFlowTasks.DAG{DataFlowTasks.DataFlowTask}, nodej::DataFlowTasks.DataFlowTask)
    @ DataFlowTasks ~/Projects/DataFlowTasks/src/dag.jl:158
  [5] addnode!(dag::DataFlowTasks.DAG{DataFlowTasks.DataFlowTask}, kv::Pair{DataFlowTasks.DataFlowTask, Tuple{Set{DataFlowTasks.DataFlowTask}, Set{DataFlowTasks.DataFlowTask}}}, check::Bool)
    @ DataFlowTasks ~/Projects/DataFlowTasks/src/dag.jl:131
  [6] addnode!
    @ ~/Projects/DataFlowTasks/src/dag.jl:120 [inlined]
  [7] addnode!
    @ ~/Projects/DataFlowTasks/src/scheduler.jl:111 [inlined]
  [8] DataFlowTasks.DataFlowTask(code::TiledFactorization.var"#6#12"{Val{false}, Matrix{Float64}, Adjoint{Float64, Matrix{Float64}}, Matrix{Float64}}, data::Tuple{Matrix{Float64}, Adjoint{Float64, Matrix{Float64}}, Matrix{Float64}}, mode::Tuple{DataFlowTasks.AccessMode, DataFlowTasks.AccessMode, DataFlowTasks.AccessMode}, priority::Int64, label::String, sch::DataFlowTasks.JuliaScheduler{DataFlowTasks.DataFlowTask})
    @ DataFlowTasks ~/Projects/DataFlowTasks/src/dataflowtask.jl:29
  [9] DataFlowTasks.DataFlowTask(code::Function, data::Tuple{Matrix{Float64}, Adjoint{Float64, Matrix{Float64}}, Matrix{Float64}}, mode::Tuple{DataFlowTasks.AccessMode, DataFlowTasks.AccessMode, DataFlowTasks.AccessMode}, priority::Int64, label::String)
    @ DataFlowTasks ~/Projects/DataFlowTasks/src/dataflowtask.jl:25
 [10] macro expansion
    @ ~/Projects/DataFlowTasks/src/dataflowtask.jl:248 [inlined]
 [11] _cholesky!(A::TiledFactorization.PseudoTiledMatrix{Float64}, tturbo::Val{false}, copy::Val{true})
    @ TiledFactorization ~/Projects/TiledFactorization/src/cholesky.jl:50
 [12] #cholesky!#2
    @ ~/Projects/TiledFactorization/src/cholesky.jl:10 [inlined]
 [13] top-level scope
    @ ~/Projects/TiledFactorization/benchmarks/cholesky_logging.jl:22
```

The issue seems to come from the fact that the *generic* `memory_overlap` method performs an `IO` operation: 

https://github.com/maltezfaria/DataFlowTasks.jl/blob/50c54abe1d14a705f892302c8d0f614f0e5f923f/src/dataflowtask.jl#L153

In fact, removing the `@warn`, the program works fine (albeit slow because many dependencies are added). I am not yet sure where the issue comes from. FWIW, perhaps it would be safer to also create a channel for inserting tasks in the `DAG`?

**Comment by maltezfaria** [comment]
Note that a temporary way out would be to instead throw an error for the generic `memory_overlap` (maybe better?), but that does not really addresses the underlying problem. Should do user overload the `memory_overlap` method and include an `IO`, the problem would still occur.

**Comment by maltezfaria** [comment]
The issue is fixed with #19 . The problem is that the `IO` in the `memory_overlap` caused the task responsible for adding a node to (sometimes) switch to the task in charge of removing nodes from the `DAG` *during* the addition of a node. Adding a lock to the `DAG`, which is now acquired/release during operations which mutate the `DAG`, fixes the issue.
